### PR TITLE
Don't use reloc shaders when generic descriptors are in the top level

### PR DESCRIPTION
**Pull request by s-perron**
The linking code cannot link the load of a resource descriptor if the
resource is in a top level user data node.  We fall back to doing a full
pipeline compilation in that case.

This is to fix the issue identified [here](https://github.com/GPUOpen-Drivers/llpc/pull/673#discussion_r430660652).

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

File: llpc/context/llpcCompiler.cpp [hunk 1]
```diff
@@ -832,11 +833,29 @@ Result Compiler::buildPipelineWithRelocatableElf(Context *context, ArrayRef<cons
...
     const ResourceMappingNode *node = nodes + i;
-    if (node->type == ResourceMappingNodeType::DescriptorTableVaPtr) {
-      if (hasUnrelocatableDescriptorNode(node->tablePtr.pNext, node->tablePtr.nodeCount))
-        return true;
-    } else if (node->type == ResourceMappingNodeType::DescriptorBufferCompact) {
+    switch (node->type) {
+    case ResourceMappingNodeType::DescriptorTableVaPtr: {
+      const ResourceMappingNode *startInnerNode = node->tablePtr.pNext;
+      const ResourceMappingNode *endInnerNode = startInnerNode + node->tablePtr.nodeCount;
+      const ResourceMappingNode *innerNode;
```
Comment (nhaehnle) [inline llpc/context/llpcCompiler.cpp:840]: Move this declaration into the loop?

Comment (s-perron) [reply llpc/context/llpcCompiler.cpp]: Sure I can do that.

File: llpc/context/llpcCompiler.cpp [hunk 2]
```diff
@@ -64,6 +64,7 @@
 #include <mutex>
 #include <set>
 #include <unordered_set>
+#include <vkgcDefs.h>
```
Comment (kuhar) [inline llpc/context/llpcCompiler.cpp:67]: nit: this is not a system header; should be `"vkgcDefs.h"`

File: llpc/context/llpcCompiler.cpp [hunk 3]
```diff
@@ -832,11 +833,28 @@ Result Compiler::buildPipelineWithRelocatableElf(Context *context, ArrayRef<cons
...
+          // The code to handle a compact descriptor cannot be easily patched, so relocatable shaders assume there are
+          // no compact descriptors.
+          return true;
+      }
+    } break;
+    case ResourceMappingNodeType::DescriptorResource:
+    case ResourceMappingNodeType::DescriptorSampler:
+    case ResourceMappingNodeType::DescriptorCombinedTexture:
+    case ResourceMappingNodeType::DescriptorTexelBuffer:
+    case ResourceMappingNodeType::DescriptorBuffer:
```
Comment (trenouf) [inline llpc/context/llpcCompiler.cpp:851]: We do need to handle buffer descriptors in the top level, to handle dynamic descriptors. See the TODO at DescBuilder.cpp:402.

File: llpc/context/llpcCompiler.cpp (hunk 2 above)
Comment (s-perron) [reply llpc/context/llpcCompiler.cpp]: Actually this header is not needed, so I just removed it.  Thanks.

File: llpc/context/llpcCompiler.cpp (hunk 3 above)
Comment (s-perron) [reply llpc/context/llpcCompiler.cpp]: Okay, I've removed buffer descriptor case.

File: llpc/context/llpcCompiler.cpp [hunk 4]
```diff
@@ -832,11 +832,27 @@ Result Compiler::buildPipelineWithRelocatableElf(Context *context, ArrayRef<cons
...
+    case ResourceMappingNodeType::DescriptorTableVaPtr: {
+      const ResourceMappingNode *startInnerNode = node->tablePtr.pNext;
+      const ResourceMappingNode *endInnerNode = startInnerNode + node->tablePtr.nodeCount;
+      for (const ResourceMappingNode *innerNode = startInnerNode; innerNode != endInnerNode; ++innerNode) {
+        if (innerNode->type == ResourceMappingNodeType::DescriptorBufferCompact)
+          // The code to handle a compact descriptor cannot be easily patched, so relocatable shaders assume there are
+          // no compact descriptors.
+          return true;
+      }
+    } break;
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:845]: Could you move the break to {} body. That is

case XXX: {
   // Code piece
  break;
}
case YYY:

Comment (s-perron) [reply llpc/context/llpcCompiler.cpp]: done.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 712849b1bd1e1510efeb9587ab596fc262e39974
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 82390674c40402c81dbf311a5ee488fb972894f5</li>
<li>SPVGEN: dev 2f679769a8491f423f5cf2f06adfa2870a82935d</li>
<li>XGL: dev 116d28a446eba3a9f8e6f577bab34fc07e090993</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev e90e9d404b64f21a69c23e2a8709e4eb56ba04e4</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx8</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Gfx9</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**